Set root component state instead of calling RenderDom for every update

alfonsogarciacaro opened the pull request:
Alternative to #6

et1975 commented:
Would you mind running performance test, with something like [this](https://github.com/et1975/react-angular-ember-elm-performance-comparison).

alfonsogarciacaro commented:
I cloned the repo but there are no instructions to build the tests. Do you know how to do it?

MangelMaxime commented:
@alfonsogarciacaro I had already added Elmish test some months ago in my fork. Please take a look at the [2 last commits](https://github.com/MangelMaxime/react-angular-ember-elm-performance-comparison).

You will probably need to update the Fable version, etc. because it's outdated but you will got the basic idea of how to add a new test.

alfonsogarciacaro commented:
Thanks @MangelMaxime! It seems I couldn't run the tests because `picker.js` was missing. I don't have Elm to compile it so I just copied it from the live example and added it to the repo. Besides that, I've [updated the elmish-react sample and added a version with the editions in this PR](https://github.com/alfonsogarciacaro/ui-perf/tree/elmish-setstate). However, the test runner fails [in this line](https://github.com/alfonsogarciacaro/ui-perf/blob/b95940ec6c654414cd2a9800d589db57c0248cfe/src/add-complete-delete.js#L66) because the DOM elements are not mounted synchronously so the buttons with classname "toggle" cannot be found. Any idea how can this be fixed? (I also tried removing `requestAnimationFrame` but that didn't work)

zaaack commented:
@alfonsogarciacaro Have you tried with input + Value ? After adding RAF `setState`+ `Value` trick didn't work in my here.

alfonsogarciacaro commented:
@zaaack Yes, [input's `Value` is working](https://github.com/alfonsogarciacaro/ui-perf/blob/b95940ec6c654414cd2a9800d589db57c0248cfe/implementations/elmish-react-setstate/todomvc.fs#L178-L186). You can try my cloning my fork and do the following:

```
cd implementations/elmish-react-setstate
yarn
dotnet restore
dotnet fable yarn-start
```

About the link to the React code, I didn't follow the references but in the comment on the top of the file it says that's a polyfill for `requestIdlCallback`, which [according to MDN](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback):

> queues a function to be called during a browser's idle periods. This enables developers to perform background and low priority work on the main event loop, without impacting latency-critical events such as animation and input response.

That makes me wonder whether this is actually used for `ReactDom.render` or `setState`, as these are hardly considered low priority work 🤔

zaaack commented:
@alfonsogarciacaro I see, I'm afraid that's mainly because you already [removed requestAnimationFrame](https://github.com/alfonsogarciacaro/ui-perf/blob/b95940ec6c654414cd2a9800d589db57c0248cfe/implementations/elmish-react-setstate/react.fs#L61-L65) I didn't look deeply into react's source code, but I think this might not be a problem, because you can easily find out by adding another ui test for it?

zaaack commented:
@alfonsogarciacaro  The input's Value didn't working in your repo after I change the raf code to:

```F#
match rootEl with
            | Some _ ->
                // observable.Trigger(model, dispatch)
                match lastRequest with
                | Some r -> window.cancelAnimationFrame r
                | None -> ()
                lastRequest <- Some (window.requestAnimationFrame (fun _ ->
                    observable.Trigger(model, dispatch)))
```

I didn't look deeply into React's source code, and cannot give some strong evidence that React using requestIdlCallback to batch setState. But I think it's not a big problem since we can easily add a new ui test case.

alfonsogarciacaro commented:
@zaaack Ah, you're right. The cursor is jumping to the end if `requestAnimationFrame` is used, didn't notice that 😅 Well, then I guess if the whole purpose of using `setState` is to avoid that glitch we'd have to remove `raf` too 😸

zaaack commented on src/react.fs:
```diff
@@ -6,21 +6,71 @@ open Fable.Core
+
     /// Setup rendering of root React component inside html element identified by placeholderId
     let withReact placeholderId (program:Elmish.Program<_,_,_,_>) =
+        let mutable rootEl = None
```
local state might break hmr, https://github.com/fable-elmish/react/pull/6/files#diff-6e208d7e773cd65ba4c7e6cf6adde006R36  maybe it's better to be put in module level?

alfonsogarciacaro replied:
Makes sense, thanks! I was also thinking that a boolean flag (like `let mutable initialized = false`) is probably enough here :+1:

zaaack replied:
I tried move the local state to global, remove `shouldComponentUpdate`, but it's strange that I cannot get HMR working with you implementation. you can try it with my demo repo by replace `paket-files/zaaack/react/src/react.fs` with yours and edit `Home/View`: https://github.com/zaaack/elmish-react-demo

alfonsogarciacaro replied:
Hi @zaaack! I just did that: cloned your demo and replaced react.fs with the [code here](https://github.com/alfonsogarciacaro/ui-perf/blob/b95940ec6c654414cd2a9800d589db57c0248cfe/implementations/elmish-react-setstate/react.fs) (also uncommented the `|> Program.withHMR` line in App.fs) and HMR worked without problem. The only change needed is to remove `shouldComponentUpdate` so I guess we can wrap it `#if !DEBUG` to keep it in Release mode. Note I didn't have to make any state global.

zaaack replied:
@alfonsogarciacaro I tried what you said before, but the view doesn't change if I edit a string child of a ReactElement, and also no errors.

alfonsogarciacaro replied:
Not sure what we're doing differently because for me it's changing.

![react-setstate](https://user-images.githubusercontent.com/8275461/33122605-fee4e0ba-cf77-11e7-9496-b32e8501bd01.gif)

However, it's true HMR doesn't work completely. Updating React elements is fine for me but if I update the logic (e.g. the incremental value in Counter/State.fs) the change is not reflected in the app unless I manually refresh (this works with the current elmish/react package).

zaaack replied:
👍  It's working now after I update to your current `react.fs`, I guess it's because the react.fs in my here is too and used a captured view:
```F#
let view = this.props.view

//...
member this.render() =
            view this.state.model this.state.dispatch
```

Now is:
```F#
member this.render() =
            this.props.view this.state.model this.state.dispatch
```

alfonsogarciacaro replied:
Any clue [about this](https://github.com/fable-elmish/react/pull/10#discussion_r152523369) @MangelMaxime?

MangelMaxime replied:
I think because, you use a stateful components at the root. If I remember, well when I did my search about React + HMR, they where having a "debug root component" and a "production root component".

Idea being, the "bug root component" was the one supporting HMR directly and knowing how to update itself from it. Like using an existing state and forcing a whole evaluation of the react components.


Current HMR implementation is really basic. It's using a private state inside the `HMR.fs` file and re-evaluate the application from scratch like if you were doing a "F5". Because the "application state" is inside the HMR.fs module then we maintain it. Then because, we use React to make a diff between a DOM "non-empty" element and a VDOM, react detect that's the DOM element is already constructed with React so it's making a diff between this two.

I remember you (@alfonsogarciacaro) provided me a link to a github sample with React + HMR. I think it's in this repo that I saw the `index.debug.js` and `index.production.js` thing.

alfonsogarciacaro replied:
Thanks a lot for the pointers @MangelMaxime! OK, [this seems to work](https://github.com/fable-elmish/react/commit/e3b56341dd1e96c861aa0d7cf57540e44505c701). What I don't understand is why _the old way_ doesn't cause the jumping cursor issue when using `Value` for an input element. Is this because of hot reloading? Or for not using `requestAnimationFrame`? Or it has been fixed in React? 🤔

MangelMaxime replied:
Don't know ..

zaaack replied:
I would vote for  `not using requestAnimationFrame`...

zaaack replied:
@alfonsogarciacaro I tried my react.fs and hmr work for update. And also yours after make local state to global state like this:

(Sorry for the downcast, I don't know how to get around `Value restriction` and just for a quick example,  so ...  😅)

```F#
open Fable.Import.Browser

    let mutable observable = None :> obj

    /// Setup rendering of root React component inside html element identified by placeholderId
    let withReact placeholderId (program:Elmish.Program<_,_,_,_>) =

        let setState model dispatch =
            match (observable :?> SingleObservable<_> option) with
            | Some obs ->
                obs.Trigger(model, dispatch)
```

alfonsogarciacaro replied:
I think I'm getting a bit lost ;) I don't see any difference by placing `observable` on the top level. But yes, I can confirm `requestAnimationFrame` is what causes the jumping cursor issue, if I add it to [these lines](https://github.com/fable-elmish/react/blob/e3b56341dd1e96c861aa0d7cf57540e44505c701/src/react.fs#L61-L63) then the problem appears again.

zaaack replied:
After I making local state to global, the new-added `printfn("update")` in update function worked, before it doesn't print anything.
![untitled3](https://user-images.githubusercontent.com/5233940/33164027-051e8cf2-d06c-11e7-85ba-46e512455b57.gif)

et1975 commented:
So, how is the perf looking?

et1975 commented:
Actually, forget performance. Can you put the old impl back in, just name it something like `withReactDOM`, for people like me who don't want the full React just for that little glitch?

alfonsogarciacaro commented:
That's a great idea @et1975! I assume keeping the old implementation means also keeping `requestAnimationFrame` (which seems to be the actual cause of the glitch), is that right?

About performance, I haven't managed yet to make this work with the test project (see [above](https://github.com/fable-elmish/react/pull/10#issuecomment-345719352)), I'll see if it can be fixed :+1:

MangelMaxime commented:
@alfonsogarciacaro I think he means to not touch the `withReact` function and create a new one called `withReactDom` which is the one you propose in this PR.

Like so users have the choice.

et1975 commented:
Either way, just as long as there's a choice.

alfonsogarciacaro commented:
Ok, I finally found why the tests didn't work, the runner is [submitting `input` events](https://github.com/alfonsogarciacaro/ui-perf/blob/b95940ec6c654414cd2a9800d589db57c0248cfe/src/add-complete-delete.js#L46) so [this event](https://github.com/alfonsogarciacaro/ui-perf/blob/b95940ec6c654414cd2a9800d589db57c0248cfe/implementations/elmish-react/todomvc.fs#L147) must be changed to `OnInput`. After doing that the tests have revealed... this PR is actually slower 😅 

<img width="845" alt="screen shot 2017-11-28 at 10 14 59 pm" src="https://user-images.githubusercontent.com/8275461/33344653-fcfaf066-d489-11e7-8535-7281f93a1cbd.png">

However, removing `requestAnimationFrame` gives better results and removes the glitch (as @zaaack has been saying all the time :wink:), should we do just that?

zaaack commented:
Thanks for your effort to fix this issue! I can confirm `requestAnimationFrame` is the reason. If removing `requestAnimationFrame` can get better performance, then I suppose [React's optimization is undering `ReactDOM.render`](https://github.com/facebook/react/blob/v15.6.2/src/renderers/dom/client/ReactMount.js#L391), and removing it is goods for everyone! 

Maybe there are some info that I forgot to mention, this "glitch" would cause asians (like me) who write CJK with IME cannot input any character, we have to using the workaround anywhere. perhaps that's why there are not much feedback about this problem..

![untitled3](https://user-images.githubusercontent.com/5233940/33354652-dfc5bb10-d4ee-11e7-9fbc-13d738906b42.gif)
![untitled4](https://user-images.githubusercontent.com/5233940/33354463-05593c54-d4ee-11e7-8f1a-9a49304db128.gif)

alfonsogarciacaro commented:
Thanks to you for such a thorough testing @zaaack I also use CJK with IME (though I hadn't tried with an Elmish app yet) so I agree this is a strong reason to fix the glitch (I also fixed [something similar](https://github.com/fable-compiler/Fable/issues/784) in Fable). What do you think @et1975? If that's ok, I will edit the PR so it just removes `requestAnimationFrame` and nothing else (well, probably also caching the location of the DOM element when React is mounted), or @zaaack can also do it in his PR :+1:

et1975 commented:
For backwards compatibility, let's have 2 functions to init react: `withReactComponent` where you can have your component instantiated and run any way you like, and `withReact` that retains my earlier implementation.

alfonsogarciacaro commented:
Sorry, I should have just updated the PR to make the intention clearer. I just did it and as you can see, the (last) proposal is to only remove `requestAnimationFrame`, no special React component is needed. Just by removing `requestAnimationFrame` the glitch is fixed.

I run the tests again, and this PR is also slower, but I guess `requestAnimationFrame` has an unfair advantage for this kind of benchmarks which will probably fade away in a normal application.

What do you think?

<img width="750" alt="screen shot 2017-11-29 at 11 01 33 pm" src="https://user-images.githubusercontent.com/8275461/33401719-869aa61a-d55a-11e7-8728-b0bcad00e94a.png">

et1975 commented:
Hmm, weird! So this must have been a regression when I introduced the RAF optimization...
I wonder how exactly it screws React, because even though it seems like it would harmless to apply this patch *I'd really like to keep the RAF*. Here's my reasoning: while a human can't type fast enough for this optimization to make a difference, something like websockets flooding with updates can easily cause so many redraws that UI appears to freeze. I wish we could pull a react dev into this to explain what's happening... Anyone got friends at Facebook? :)
Failing that... let's think this through... RAF should make no difference when a human is typing, right? And yet it does? So what's different?

MangelMaxime commented:
I am a bit loss here. In the latest state of this PR what is it trying to solve ?

Because, at first it was about using `setState` ? But now, it's about removing the RAF ?

*Aparté
Also, in the test benchmark Elm is not using RAF. [Source](https://github.com/evancz/react-angular-ember-elm-performance-comparison#no-requestanimationframe-in-elm) so the benchmark should not use it either.*

alfonsogarciacaro commented:
@MangelMaxime Yes, this and #6 where originally intended to use `setState` in the root, because we thought this was causing the glitch. But with the tests and thanks to @zaaack's help, we realize the real issue was the `requestAnimationFrame` optimization, that's why I changed the PR.

 > Yes, Elms is not using `requestAnimationFrame` in the tests so it doesn't have an _unfair_ advantage, and we do, but that's another topic :wink:

@et1975 Following your previous suggestion, I've modified the PR to keep the original implementation and offer an alternative with the explanation. I also opened #12 to keep better track of the issue and try to investigate the real cause of the problem :+1:

et1975 commented:
I've released this as 1.0.1-beta-1, please hard reset on origin/master as I've squashed all those commits to keep the history clean.
